src/mono/mono/metadata/threads.c

```diff
@@ -1266,9 +1275,16 @@ start_wrapper (gpointer data)
 	info = mono_thread_info_attach ();
 	info->runtime_thread = TRUE;
 
+	gboolean external_eventloop = start_info->external_eventloop;
 	/* Run the actual main function of the thread */
 	res = start_wrapper_internal (start_info, (gsize*)info->stack_end);
 
+	if (G_UNLIKELY (external_eventloop)) {
```

Comment:
How will this thread be used after it has returned from its thread function? Will we use it to run other code (JS event loop), "impersonating" the thread, like having events that should run managed code, referencing this thread, so when run picked up in JS even loop, it will still be able to complete those events, even if the original thread execution has stopped?

Reply:
Yes, exactly right.

In order for JS promises to resolve, the JS eventloop must get a chance to run.  So if we create a worker thread with `Task.Run` and  `await` on an imported JS promise (which gets exposed as a `Task<T>` in C#) the worker thread must be able to run the JS eventloop and then invoke the managed code once the JS promise is finished.

The earlier PRs (see the PR description) refactored the managed threadpool worker code to use callbacks that are registered with the JS event loop in order to signal a threadpool worker when it has work available.  The worker stays alive as long as there are unsettled JS promises (that are awaited by managed code) and then either exits (normal `pthread_exit` shuts down these "external eventloop" pthreads same as normal pthreads) or picks up additional work - depending on what the threadpool algorithm says.

Reply:
> If so, will the thread move over to safe mode since it will still be attached to runtime, but not able to react to any suspend requests etc?

yes it should be in GC Safe when it's in the JS loop (I'll do a checked build this afternoon to verify.  but I'm pretty sure that actually happens as part of the normal thread state management for the managed thread start function invocation. when it returns we will go to gc safe)

Reply:
> In order for JS promises to resolve, the JS eventloop must get a chance to run. So if we create a worker thread with `Task.Run` and `await` on an imported JS promise (which gets exposed as a `Task<T>` in C#) the worker thread must be able to run the JS eventloop and then invoke the managed code once the JS promise is finished.

So then we keep a reference count or similar making sure the managed thread gets detached when there are no more outstanding references to the thread, correct? Sounded like we depend on TLS destructors to do that, so does code reacting to callback from JS eventloop, invoke managed code and then check refcount and if that is 0, close native thread, that in turn will trigger TLS destructors, detaching thread from runtime?

Reply:
Correct

Reply:
Added a gc safe assertion when we're exiting to the external event loop